### PR TITLE
Rspamd: more features

### LINKED ISSUE #3145: [FR] Rspamd: learn from moving from/to junk folder
# Feature Request

## Context

Rspamd is able to learn ham/spam by moving mails from/to the junk folder. This is a nice addition.

### Is your Feature Request related to a Problem?

No.

### Describe the Solution you'd like

See <https://www.kernel-error.de/2020/05/04/rspamd-automatisch-spam-ham-lernen-mit-dovecot-und-imapsieve/> (German though).

### Are you going to implement it?

Yes, because _I always do it anyway_.

### Who will that Feature be useful to?

Everyone using Rspamd.

### What have you done already?

Nothing.

### LINKED ISSUE #3156: [FR] Rspamd: make it possible to move e-mails to Junk automatically
# Feature Request

## Context

Users should be able to toggle behavior to automatically move mail to the Junk folder when Rspamd adds the `X-Spam: Yes` header.

### Is your Feature Request related to a Problem?

Currently, users must move them automatically.

### Describe the Solution you'd like

This should be done by Sieve automatically if desired.

### Are you going to implement it?

Yes.

### What are you going to contribute??

Everything.

### PR DESCRIPTION
**Pull request by georglauterbach**
# Description

Add new features to our Rspamd integration:

1. With `MOVE_SPAM_TO_JUNK=1`, we will now move messages marked with the `X-Spam: Yes` header to the "Junk" folder
2. With `RSPAMD_LEARN=1`, we enable
   1. auto-learning for the Bayes filter
   2. when a user moves messages from or to the "Junk" folder, Rspamd will learn them as ham or spam (for the Bayes filter as well)

I tried to separate commits as cleanly as possible to make reviews easier.

Closes #3145
Closes #3156

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by georglauterbach** [comment]
@hanscees @alexanderadam you could really help me here: do you know how to trigger Sieve locally by moving a mail into the "Junk" folder with a command executed inside the container? I want to complete the last test for `RSPAMD_LEARN`.

**Comment by alexanderadam** [comment]
> do you know how to trigger Sieve locally by moving a mail into the "Junk" folder with a command executed inside the container?

Sadly I don't 😬 
But I would guess copying a SPAM mail into the `new` folder maybe?
Sorry, but I'm not very good with this email infrastructure thing.

**Comment by hanscees** [comment]
I have to think about that. Will come back to it this evening. 
You basically have to log in as imap user. Then do a move from folder a to folder b.

**Comment by georglauterbach** [comment]
> I have to think about that. Will come back to it this evening. 

Thanks!

> You basically have to log in as imap user. Then do a move from folder a to folder b. 

Indeed. I don't know how to do that actually, I have never done that before.

**Comment by casperklein** [comment]
> @polarathene @casperklein would you be fine if this PR got included into v12.0.0? I marked it for v12.1.0, but if you're okay with it, I'd take care of this ASAP so we can merge this for v12.0.0.

Why shouldn't it be fine to include it in v12? I see no reason against it.

File: target/scripts/startup/setup.d/security/rspamd.sh
Comment (casperklein) [inline target/scripts/startup/setup.d/security/rspamd.sh:15]:
```suggestion
    [[ ${RSPAMD_MOVE_SPAM_TO_JUNK} -eq 1 ]] && __rspamd__setup_move_spam_to_junk_automatically
```

File: target/scripts/startup/setup.d/security/rspamd.sh
Comment (casperklein) [inline target/scripts/startup/setup.d/security/rspamd.sh:263]: I guess a file change here is expected. If so, I advice to consequently use `sedfile -i` instead of `sed -i`

**Comment by georglauterbach** [comment]
> I am not familiar with dockers testing, but I assume you can spin up an image and run commands in it.

That's what we do :)


> the notes to do copy's with imap are here: see https://www.atmail.com/blog/advanced-imap/

Thanks for the link, will have a look.

> You would have to make a user. Send email in a folder and then move it. Then look in the logging of dovecot.

The Rspamd test sends 4 emails actually: 1 normal, 1 Gtube rejection, one virus rejection, and one that Rspamds adds a header to.

The part that moves the mail does not trigger Sieve for me though.

> And you would need some sort of rather large expect script. Perhaps these do exists, don't know.

Don't worry, our test suite (BATS) plus our helpers are rather comprehensive.


> Perhaps more manageable is to run and parse the command doveconf and look for the plugin section and check a couple of things :

Test is already doing that.

> Does this help a bit? 

To will have a look at the link you provided. The rest is already all implemented :)

**Comment by georglauterbach** [comment]
> Why shouldn't it be fine to include it in v12? I see no reason against it.

👍🏼

**Comment by georglauterbach** [comment]
@casperklein @polarathene you can already go ahead and review changes. The only additions I will try to implement is for the last test to be more comprehensive.

**Comment by polarathene** [comment]
> I marked it for v12.1.0, but if you're okay with it, I'd take care of this ASAP so we can merge this for v12.0.0.

This is probably going to be low on my priority to get to. Or if @casperklein reviews before me and approves, that's fine to merge without waiting on my review input :+1: 

I am probably going to need a day or so before I could make time for it so by all means if that suits you go for it.

**Comment by casperklein** [comment]
Generally, "move_spam_to_junk" isn't a spamassassin/rspamd specific thing.

I think there is no need for a separate `RSPAMD_MOVE_SPAM_TO_JUNK` ENV / function.

https://github.com/docker-mailserver/docker-mailserver/blob/dab70709d90df1d83b920eccc6c0386b283d677e/target/scripts/startup/setup.d/dovecot.sh#L77-L85

Just remove the `${SPAMASSASSIN_SPAM_TO_INBOX} -eq 1` condition and `MOVE_SPAM_TO_JUNK` can be used universally.

**Comment by georglauterbach** [comment]
@casperklein you are right! I went ahead and implemented your approach, then rewrote history to have cleanly separated commits again; suggestions are incorporated :)

I will force-push my changes now.

**Comment by georglauterbach** [comment]
FYI: I solved the issue about not being able to trigger Sieve with `doveadm` by not using `doveadm` but by directly talking IMAP4 over netcat :D Works like a charm!

**Comment by georglauterbach** [comment]
I also apologize for not making this two PRs. Initially, changes were not supposed to be this big. But the Sieve adjustments (I deem necessary) are about 100 SLoC. The commits are cleanly separated though to make reviewing easier.

**Comment by georglauterbach** [comment]
FYI: I'd like to merge this at the very latest on Friday, 12 o'clock UTC+1 because thereafter I will be unavailable to merge/adjust for 2 weeks straight.

**Comment by polarathene** [comment]
> Friday, 12 o'clock UTC+1

12 hours to go :sweat_smile: 

I'll try to review before then, but if another maintainer approves that's fine by me too for merging.

File: target/scripts/startup/setup.d/security/misc.sh
Comment (polarathene) [inline target/scripts/startup/setup.d/security/misc.sh:282]: Is it not relevant to also `rm` the `.svbin`? This seemed to be done previously.

```suggestion
    rm /usr/lib/dovecot/sieve-global/after/spam_to_junk.{sieve,svbin}
```

File: target/scripts/startup/setup.d/dovecot.sh
Comment (polarathene) [inline target/scripts/startup/setup.d/dovecot.sh:55]: Just a note that there is no more `rm -f` cleanup on these locations if they contain anything (_not that they should, but was part of previous restart cleanup behaviour I assume_).

Likewise for `.sieve` and `.svbin` files in `sieve-global/{before,after}`.

**Review by polarathene — APPROVED** [review]
Light review as short on time and Dovecot sieve + rspamd aren't anything I am too knowledgeable about.

I've left some notes, but if they're not a concern in your opinion that's fine, go ahead and merge :+1: Happy vacation! :grinning:

**Comment by casperklein** [comment]
> FYI: I'd like to merge this at the very latest on Friday, 12 o'clock UTC+1 because thereafter I will be unavailable to merge/adjust for 2 weeks straight.

I wasn't able to review yesterday :-\ Tonight/tomorrow I can review.

File: target/scripts/startup/setup.d/security/misc.sh
Comment (georglauterbach) [reply target/scripts/startup/setup.d/security/misc.sh:282]: The `.svbin` file does not exists, because it is created only in case the `if` evaluated to true and `sievec` is called.

File: target/scripts/startup/setup.d/dovecot.sh
Comment (georglauterbach) [reply target/scripts/startup/setup.d/dovecot.sh:88]: The primary reason for removal was that the `chmod` affected files and directories, but they need separate privilges.

File: target/scripts/startup/setup.d/dovecot.sh
Comment (georglauterbach) [reply target/scripts/startup/setup.d/dovecot.sh:90]: The dirst are fine with `rwx` for Dovecot and root; and other can list and enter them. I think I tested `XX0` previously but had weird permission issues. The `775` value (probably `755` works as well, I will test that) is an empirical value really.

Comment (georglauterbach) [reply target/scripts/startup/setup.d/dovecot.sh:90]: The change to `dovecot:root` I had seen in a tutorial and it made absolutely sense - `docker:docker` does not - at all. The permissions were somehow pretty messed up before, and we still have a `chmod +x` for all files in the Sieve pipe binary directory, which I find odd, but couldn't get it to work without it.

File: target/scripts/startup/setup.d/security/misc.sh
Comment (georglauterbach) [reply target/scripts/startup/setup.d/security/misc.sh:274]: Good catch for the `dovecot:root`! I will try to fix that. The `+x` is not needed though.

File: target/scripts/startup/setup.d/dovecot.sh
Comment (georglauterbach) [reply target/scripts/startup/setup.d/dovecot.sh:55]: That is intentional. I found no good reason for deleting all files there. And it works flawlessly without. I found that the dirs are empty to begin with, so the cleanup is irrelevant.

**Comment by georglauterbach** [comment]
> 

I won't be able to properly respond to feedback then; I got about 3 hours left for that, as I said.

**Comment by georglauterbach** [comment]
I was able to feedback from @polarathene now; I hope you like the PR @casperklein :)

Thanks for the nice vacation wish

> Happy vacation! 😀

Thank's a lot! :D

File: target/scripts/startup/setup.d/security/misc.sh
Comment (polarathene) [inline target/scripts/startup/setup.d/security/misc.sh:280]: Honestly not a fan of `EOF` like this vs separate files, but I've said this before with cron jobs and looked the other way in the PR where this was also done :sweat_smile: 

Separate config files have better observability in my opinion.
- Far more visible in the file tree as distinct content from scripts. Also beneficial for tasks like `grep` on content that isn't `.sh` when searching / filtering the entire project for something, less noise to wade through.
- `git blame` far more useful as they don't tend to go through as much churn as scripts have been (_each year it seems they get refactored heavily_).

I don't think there's much gain from inlining content like this, but I know it's preferred by you and casper :sweat_smile:

**Review by polarathene — APPROVED** [review]
I suppose we merge this after your feedback @casperklein as @georglauterbach is no longer able to respond for 2 weeks.

Feel free to merge after your review or make any changes :+1:

**Comment by georglauterbach** [comment]
> I suppose we merge this after your feedback @casperklein as @georglauterbach is no longer able to respond for 2 weeks.

Very quick feedback I can give, will be on GH for 1 minute a day :) Flight hasn't yet started :D I just can't edit or adjust anything

> Feel free to merge after your review or make any changes :+1: 

Yes - go ahead! :)

File: target/scripts/startup/setup.d/security/misc.sh
Comment (georglauterbach) [reply target/scripts/startup/setup.d/security/misc.sh:280]: I know you're not a fan, but from my perspective, the gains (reduced complexity in the file tree, Dockerfile and removal of the uncanny `rm` in the `else` branch) outweigh the disadvantages in this case for such a small file :) 

With larger files, I agree with you!

**Comment by georglauterbach** [comment]
> > FYI: I'd like to merge this at the very latest on Friday, 12 o'clock UTC+1 because thereafter I will be unavailable to merge/adjust for 2 weeks straight.
> 
> I wasn't able to review yesterday :-\ Tonight/tomorrow I can review.

ping :)

**Comment by casperklein** [comment]
> > I wasn't able to review yesterday :-\ Tonight/tomorrow I can review.
> 
> ping :)

1. Enjoy your vacation and don't hang around at github 😝
2. Sorry for the delay again, I am currently out of order.. 🤒

File: target/scripts/startup/setup.d/security/misc.sh
Comment (casperklein) [reply target/scripts/startup/setup.d/security/misc.sh:280]: > I don't think there's much gain from inlining content like this, but I know it's preferred by you and casper 😅

I depends on 😆 For static content, I generally prefer plain files. For small snippets (with variables), a heredoc may better suited.
If there is consent, we can in a separate PR move some (not only rspam related) heredocs to files.

**Review by casperklein — APPROVED** [review]
Finally 🎉

File: target/scripts/startup/setup.d/security/misc.sh
Comment (georglauterbach) [reply target/scripts/startup/setup.d/security/misc.sh:280]: Agreed!

Comment (polarathene) [reply target/scripts/startup/setup.d/security/misc.sh:280]: > If there is consent, we can in a separate PR move some (not only rspam related) heredocs to files.

I can live with it :sweat_smile: 

There's some merit for the co-location to that extent perhaps, but those less familiar where everything is and wanting to contribute or troubleshoot something may first expect files in the container that we add to exist as files in our project, rather than embedded into scripts :man_shrugging: (_I know this was a bit of a pain-point for me with the cron jobs, which I think existed in several locations in the container, some from packages beyond what DMS added_)

It probably won't make much of a difference to getting new maintainers onboard though, so if it's easier for those actively maintaining the project to leverage heredocs, go for it :sunglasses:

**Comment by HeySora** [comment]
@georglauterbach Commit https://github.com/docker-mailserver/docker-mailserver/pull/3159/commits/b918de1381fd93bbf71985a4c27c64095faa6d64 introduced a breaking change:

> The sieve rules are now global "after" rules; this makes more sense than "before" rules..

Having the rules as "before" rules allowed users to manually "whitelist" a sender whose e-mails could be wrongly flagged as spam (with a `fileinto` to the inbox), since the user rules were parsed after.  
With this change, this is not possible anymore, as this sieve rule will now override any user filter, thus making user whitelists impossible.

May I ask what was the reasoning behind this change? I'm asking because I might be missing the point. Thanks in advance!

**Comment by georglauterbach** [comment]
I genuinly didn't think the _last_ `fileinto` directive is the one that is used... now all makes sense. I can try to fix this up and provide a PR.

Thanks a lot for the explanation!

**Comment by HeySora** [comment]
.....Actually I think I just got hit with the syndrome of "asking for help / telling someone something, to then find an answer by yourself" 😅 Apologies!

It might be worth trying to keep the Junk filter as an "after" filter just like you did, and attempt using `stop;` from user filters to see if that would work? That might actually be the right call instead.

I'm just not sure if a `stop;` only halts the execution of the current (here, user) script, or if it halts everything, including "after" filters.

**Comment by georglauterbach** [comment]
Haha, I know this "syndrom" very well :D 

What I found about your question: <https://dovecot.org/list/dovecot/2018-March/111384.html> do it is possible.

I am uncertain which route to go down. I will adjust the release notes though to make users aware of this unintended change.

**Comment by georglauterbach** [comment]
Please reach back to me whether you think the `stop;` is the right approach or whether we should return to the old way. I do prefer the `stop;` (disabling implicit keep) though, as it seems to be the cleaner solution to me.

**Comment by HeySora** [comment]
> Please reach back to me whether you think the `stop;` is the right approach or whether we should return to the old way. I do prefer the `stop;` (disabling implicit keep) though, as it seems to be the cleaner solution to me.

Hi @georglauterbach, I can now confirm after a while that this has been working flawlessly! 😄

**Pull request merged**
